File: app-server/src/main/java/com/github/arhor/simple/expense/tracker/web/validation/DateRangeValidator.java
package com.github.arhor.simple.expense.tracker.web.validation;

import javax.validation.ConstraintValidator;
import javax.validation.ConstraintValidatorContext;

import com.github.arhor.simple.expense.tracker.service.DateRangeCriteria;

public class DateRangeValidator implements ConstraintValidator<ValidDateRange, DateRangeCriteria> {

    @Override
    public boolean isValid(final DateRangeCriteria value, final ConstraintValidatorContext context) {
        if (value != null) {
            var startDate = value.startDate();
            var endDate = value.endDate();

            if ((startDate != null) && (endDate != null)) {
                return startDate.isBefore(endDate)
                    || startDate.isEqual(endDate);
            }
        }
        return true;
    }
}

File: app-server/src/main/kotlin/com/github/arhor/simple/expense/tracker/web/validation/DateRangeValidator.kt
package com.github.arhor.simple.expense.tracker.web.validation

import com.github.arhor.simple.expense.tracker.web.DateRangeCriteria
import jakarta.validation.ConstraintValidator
import jakarta.validation.ConstraintValidatorContext

class DateRangeValidator : ConstraintValidator<ValidDateRange, DateRangeCriteria> {

    override fun isValid(value: DateRangeCriteria?, context: ConstraintValidatorContext?): Boolean {
        value?.let { (startDate, endDate) ->
            if ((startDate != null) && (endDate != null)) {
                return startDate.isBefore(endDate)
                    || startDate.isEqual(endDate)
            }
        }
        return true
    }
}
